catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,422 @@
+            sourceKey);
+      }
+
+      // Kick off a thread to find a new value
+      LOGGER.trace("Starting a process to get the {} for source {}", valueName, sourceKey);
+      final ListenableFuture getNewValueFuture =
+          Futures.withTimeout(
+              // TODO Handle RejectedExecutionException?
+              pollThreadPool.submit(
+                  () -> {
+                    try {
+                      final T newValue = getValueFunction.apply(source);
+
+                      // TODO Is there a simpler way to cache the value as soon as we know
+                      // it (e.g. handling futures by first completed)?
+                      LOGGER.trace(
+                          "Successfully found {}={} for source {}", valueName, newValue, sourceKey);
+                      cacheNewValue(sourceKey, optionalOldValue, newValue);
+                    } catch (RuntimeException e) {
+                      LOGGER.debug(
+                          "RuntimeException getting the {} for source {}", valueName, sourceKey, e);
+
+                      if (optionalValueOnException.isPresent()) {
+                        final T valueOnException = optionalValueOnException.get();
+                        cacheNewValue(sourceKey, optionalOldValue, valueOnException);
+                      } else {
+                        LOGGER.debug(
+                            "Not updating the cache for the RuntimeException encountered when getting the {} for source {} because the optionalValueOnException is empty",
+                            valueName,
+                            sourceKey);
+                      }
+                    }
+                  }),
+              timeout,
+              timeoutTimeUnit,
+              pollTimeoutWatcherThreadPool);
+
+      futureSourceKeyAndOptionalOldValueMap.put(
+          getNewValueFuture, new ImmutablePair<>(sourceKey, optionalOldValue));
+    }
+    LOGGER.trace(
+        "Successfully started a process to get the {} for each of the {} current sources",
+        valueName,
+        currentSourcesCount);
+
+    // Now go through each of the Futures one-by-one. If the Future completes successfully without
+    // throwing an Exception, the new value was already successfully cached. If an Exception is
+    // thrown, check if it is a TimeoutException. In that case, cache the valueOnTimeout. Otherwise,
+    // there is some other unexpected (thread-related) error so do not update the cache.
+    for (final Map.Entry<Future, Pair<SourceKey, Optional<T>>> futureSourceKeyAndOldValueEntry :
+        futureSourceKeyAndOptionalOldValueMap.entrySet()) {
+      final Future future = futureSourceKeyAndOldValueEntry.getKey();
+      final SourceKey sourceKey = futureSourceKeyAndOldValueEntry.getValue().getLeft();
+      final Optional<T> optionalOldValue = futureSourceKeyAndOldValueEntry.getValue().getRight();
+
+      try {
+        future.get(); // TODO Would it be safe to still add a timeout here?
```

Comment:
❓

Reply:
What do you mean by adding a timeout here? Doesn't the future already timeout on it's own?

Reply:
The future does already have a timeout. I am asking if we should add another timeout to be more safe, i.e. `future.get(timeout, timeoutTimeUnit);` or if we can rely on the com.google.common.util.concurrent.Futures.withTimeout Future to always finish.

Reply:
I think adding a timeout here would just be confusing since you are already handling the timeouts up above